compiler/rustc_hir_typeck/src/_if.rs

```diff
@@ -0,0 +1,281 @@
+                return (chain, IfChainTail::Missing(current_if));
+            };
+
+            if let hir::ExprKind::If(..) = else_expr.kind {
+                current_if = else_expr;
+                continue;
+            }
+
+            return (chain, IfChainTail::FinalElse(self.check_branch_body(else_expr, expected)));
+        }
+    }
+
+    fn collect_if_branch(
+        &self,
+        if_expr: &'tcx hir::Expr<'tcx>,
+        expected: Expectation<'tcx>,
+        chain: &mut IfGuardedBranches<'tcx>,
+    ) -> Option<&'tcx hir::Expr<'tcx>> {
+        let hir::ExprKind::If(cond_expr, then_expr, opt_else_expr) = if_expr.kind else {
+            bug!("expected `if` expression, found {:#?}", if_expr);
+        };
+
+        let (cond_ty, cond_diverges) = self.check_if_condition(cond_expr, then_expr.span);
+        if let Err(guar) = cond_ty.error_reported() {
+            chain.cond_error.get_or_insert(guar);
+        }
+        let branch_body = self.check_branch_body(then_expr, expected);
+
+        chain.branches.push(IfGuardedBranch { if_expr, cond_diverges, body: branch_body });
+
+        opt_else_expr
+    }
+
+    fn reset_diverges_to_maybe(&self) {
+        self.diverges.set(Diverges::Maybe);
+    }
+
+    fn take_diverges(&self) -> Diverges {
+        let diverges = self.diverges.get();
+        self.reset_diverges_to_maybe();
+        diverges
+    }
+
+    fn ensure_if_branch_type(&self, hir_id: HirId, ty: Ty<'tcx>) {
+        let mut typeck = self.typeck_results.borrow_mut();
+        let mut node_ty = typeck.node_types_mut();
+        node_ty.entry(hir_id).or_insert(ty);
+    }
+
+    fn overwrite_if_branch_type(&self, hir_id: HirId, ty: Ty<'tcx>) {
+        let mut typeck = self.typeck_results.borrow_mut();
+        let mut node_ty = typeck.node_types_mut();
+        node_ty.insert(hir_id, ty);
+    }
+
+    fn check_branch_body(
+        &self,
+        expr: &'tcx hir::Expr<'tcx>,
+        expected: Expectation<'tcx>,
+    ) -> BranchBody<'tcx> {
+        self.reset_diverges_to_maybe();
+        let ty = self.check_expr_with_expectation(expr, expected);
+        let diverges = self.take_diverges();
+        let span = self.find_block_span_from_hir_id(expr.hir_id);
+        BranchBody { expr, ty, diverges, span }
+    }
+
+    fn explain_if_branch_mismatch(&self, err: &mut Diag<'_>, prev_branch_span: Option<Span>) {
+        if let Some(prev_branch_span) = prev_branch_span {
+            err.span_label(prev_branch_span, "expected because of this");
+        }
+    }
+
+    fn set_diverges(
+        &self,
+        initial_diverges: Diverges,
+        guarded: &IfGuardedBranches<'tcx>,
+        tail: &IfChainTail<'tcx>,
+    ) {
+        let mut tail_diverges = tail.diverges();
+        for branch in guarded.branches.iter().rev() {
+            tail_diverges = branch.cond_diverges | (branch.body.diverges & tail_diverges);
+        }
+        self.diverges.set(initial_diverges | tail_diverges);
+    }
+}
```

Comment:
I'm actually not a fan of splitting this (along with `reset_diverges_to_maybe` and `take_diverges`) into separate functions.

I guess, the more I look over this, I also am not a huge fan of `explain_if_branch_mismatch` either.